adb/adb_commands.py

```diff
@@ -54,9 +54,51 @@ class AdbCommands(object):
   protocol_handler = adb_protocol.AdbMessage
   filesync_handler = filesync_protocol.FilesyncProtocol
 
-  @classmethod
-  def ConnectDevice(
-      cls, port_path=None, serial=None, default_timeout_ms=None, **kwargs):
+  def __init__(self):
+
+    self.__reset()
+
+  def __reset(self):
+    self.handle = None
+    self._device_state = None
+
+    # Connection table tracks each open AdbConnection objects per service type
+    # By default, the only service connections that make sense to hold open is (interactive) shell
+    self._service_connections = {
+        b'shell:': None
```

Comment:
Okay, I've read more and it looks like `_get_service_connection` is only used by `shell:` connections, so the logic in it is repeated. How about adding a `cache=False` kwarg to it and using it everywhere `self.protocol_handler.Open()` is called?

Reply:
This becomes a lot more useful for another branch I'm working on right now for implementing the 'adb forward' functionality. There, we'll need to hold/cache the service connection for our proxy while the TCP connection is open. I'll push that up when we're happy with this service connections implementation

Having a cache param in Open() isn't a bad idea,  but since the Protocol Handler (AdbMessage) is basically a static class, it cannot hold state data like our get_service_connections method within AdbCommands.

I could refactor calls to protocol_handler.Open() and make them pass through a CreateServiceConnection() sort of method, with a cache=False parameter. Or, leave it to the individual program functions to handle it on their own (as it is now). Fine with either way, let me know what you think.

Reply:
I've also updated the docs around self._service_connections so its a bit more clear that any service that needs to be persisted can take advantage of it, not just InteractiveShell